# Importing Libraries

In [43]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [44]:
path = r'C:\Users\tee3a\3D Objects\12-2022 Instacart Basket Analysis'

In [45]:
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data for storing any manipulated data sets', 'ords_prods_merge_new columns.pkl'))

In [46]:
# Create a subset

df = ords_prods_merge[:1000000]

In [47]:
df.shape

(1000000, 19)

In [48]:
df.head(10)

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   
5           1  Chocolate Sandwich Cookies        61             19     5.8   
6           1  Chocolate Sandwich Cookies        61             19     5.8   
7           1  Chocolate Sandwich Cookies        61             19     5.8   
8           1  Chocolate Sandwich Cookies        61             19     5.8   
9           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id eval_set  order_number  orders_day_of_week  \
0   3139998      138    prior            28                   6   
1   1977647      138    prior            30                   6   
2    389851      709    prior             2                   0   
3    652770      764    prior             1                   3   
4   1813452      764    prior             3                   4   
5   1701441      777    prior            16                   1   
6   1871483      825    prior             3                   2   
7   1290456      910    prior            12                   3   
8    369558     1052    prior            10                   1   
9    589712     1052    prior            15                   1   

   order_hour_of_day  days_since_prior_order  add_to_cart_order  reordered  \
0                 11                3.000000                  5          0   
1                 17               20.000000                  1          1   
2                 21                6.000000                 20          0   
3                 13                1.111484                 10          0   
4                 17                9.000000                 11          1   
5                  7               26.000000                  7          0   
6                 14               30.000000                  2          0   
7                 10               30.000000                  1          0   
8                 20               19.000000                  1          0   
9                 12               15.000000                  2          1   

  _merge    price_range_loc     busiest_day  Busiest_days  \
0   both  Mid-range product  Regularly busy  Regular days   
1   both  Mid-range product  Regularly busy  Regular days   
2   both  Mid-range product     Busiest day  Busiest days   
3   both  Mid-range product  Regularly busy  Slowest days   
4   both  Mid-range product      Least busy  Slowest days   
5   both  Mid-range product  Regularly busy  Busiest days   
6   both  Mid-range product  Regularly busy  Regular days   
7   both  Mid-range product  Regularly busy  Slowest days   
8   both  Mid-range product  Regularly busy  Busiest days   
9   both  Mid-range product  Regularly busy  Busiest days   

  busiest_period_of_day  
0        Average orders  
1        Average orders  
2        Average orders  
3        Average orders  
4        Average orders  
5        Average orders  
6        Average orders  
7           Most orders  
8        Average orders  
9        Average orders

In [49]:
df.groupby('product_name')

In [50]:
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.577493
2                17.320781
3                16.084944
4                17.530458
5                14.763075
6                16.658449
7                17.031590
8                15.076662
9                15.447580
10               18.681852
11               15.447411
12               14.327957
13               16.548642
14               16.960241
15               16.121948
16               17.803851
17               15.593633
18               19.674252
19               16.899756
20               16.255442
21               25.535479

In [51]:
df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max' , 'median']})

order_number               
                      mean min max median
department_id                            
1                15.577493   1  99   10.0
2                17.320781   1  96   11.0
3                16.084944   1  99   10.0
4                17.530458   1  99   11.0
5                14.763075   1  99    9.0
6                16.658449   1  99   10.0
7                17.031590   1  99   11.0
8                15.076662   1  98   10.0
9                15.447580   1  99   10.0
10               18.681852   1  99   12.0
11               15.447411   1  99   10.0
12               14.327957   1  99    9.0
13               16.548642   1  99   10.0
14               16.960241   1  99   11.0
15               16.121948   1  99   10.0
16               17.803851   1  99   11.0
17               15.593633   1  99   10.0
18               19.674252   1  99   14.0
19               16.899756   1  99   11.0
20               16.255442   1  99   10.0
21               25.535479   1  98   19.0

# Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

In [52]:
# Step 1

ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [53]:
ords_prods_merge.head(10)

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   
5           1  Chocolate Sandwich Cookies        61             19     5.8   
6           1  Chocolate Sandwich Cookies        61             19     5.8   
7           1  Chocolate Sandwich Cookies        61             19     5.8   
8           1  Chocolate Sandwich Cookies        61             19     5.8   
9           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id eval_set  order_number  orders_day_of_week  \
0   3139998      138    prior            28                   6   
1   1977647      138    prior            30                   6   
2    389851      709    prior             2                   0   
3    652770      764    prior             1                   3   
4   1813452      764    prior             3                   4   
5   1701441      777    prior            16                   1   
6   1871483      825    prior             3                   2   
7   1290456      910    prior            12                   3   
8    369558     1052    prior            10                   1   
9    589712     1052    prior            15                   1   

   order_hour_of_day  days_since_prior_order  add_to_cart_order  reordered  \
0                 11                3.000000                  5          0   
1                 17               20.000000                  1          1   
2                 21                6.000000                 20          0   
3                 13                1.111484                 10          0   
4                 17                9.000000                 11          1   
5                  7               26.000000                  7          0   
6                 14               30.000000                  2          0   
7                 10               30.000000                  1          0   
8                 20               19.000000                  1          0   
9                 12               15.000000                  2          1   

  _merge    price_range_loc     busiest_day  Busiest_days  \
0   both  Mid-range product  Regularly busy  Regular days   
1   both  Mid-range product  Regularly busy  Regular days   
2   both  Mid-range product     Busiest day  Busiest days   
3   both  Mid-range product  Regularly busy  Slowest days   
4   both  Mid-range product      Least busy  Slowest days   
5   both  Mid-range product  Regularly busy  Busiest days   
6   both  Mid-range product  Regularly busy  Regular days   
7   both  Mid-range product  Regularly busy  Slowest days   
8   both  Mid-range product  Regularly busy  Busiest days   
9   both  Mid-range product  Regularly busy  Busiest days   

  busiest_period_of_day  max_order  
0        Average orders         32  
1        Average orders         32  
2        Average orders          5  
3        Average orders          3  
4        Average orders          3  
5        Average orders         26  
6        Average orders          9  
7           Most orders         12  
8        Average orders         20  
9        Average orders         20

In [54]:
# Step 2

ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [55]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [56]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [57]:
# loyalty_flag frequency

ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

In [58]:
# Step 3

ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(10)

user_id      loyalty_flag  order_number
0      138  Regular customer            28
1      138  Regular customer            30
2      709      New customer             2
3      764      New customer             1
4      764      New customer             3
5      777  Regular customer            16
6      825      New customer             3
7      910  Regular customer            12
8     1052  Regular customer            10
9     1052  Regular customer            15

#  What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [59]:
ords_prods_merge[['prices', 'loyalty_flag']].head(10)
                                                

prices      loyalty_flag
0     5.8  Regular customer
1     5.8  Regular customer
2     5.8      New customer
3     5.8      New customer
4     5.8      New customer
5     5.8  Regular customer
6     5.8      New customer
7     5.8  Regular customer
8     5.8  Regular customer
9     5.8  Regular customer

It was noted from the above data that the same price that the regular customers get their orders is the same for new and loyal customers.

# task 4.8

In [60]:
ords_prods_merge.shape

(32404859, 21)

In [61]:
ords_prods_merge.head(10)

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   
5           1  Chocolate Sandwich Cookies        61             19     5.8   
6           1  Chocolate Sandwich Cookies        61             19     5.8   
7           1  Chocolate Sandwich Cookies        61             19     5.8   
8           1  Chocolate Sandwich Cookies        61             19     5.8   
9           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id eval_set  order_number  orders_day_of_week  ...  \
0   3139998      138    prior            28                   6  ...   
1   1977647      138    prior            30                   6  ...   
2    389851      709    prior             2                   0  ...   
3    652770      764    prior             1                   3  ...   
4   1813452      764    prior             3                   4  ...   
5   1701441      777    prior            16                   1  ...   
6   1871483      825    prior             3                   2  ...   
7   1290456      910    prior            12                   3  ...   
8    369558     1052    prior            10                   1  ...   
9    589712     1052    prior            15                   1  ...   

   days_since_prior_order  add_to_cart_order  reordered  _merge  \
0                3.000000                  5          0    both   
1               20.000000                  1          1    both   
2                6.000000                 20          0    both   
3                1.111484                 10          0    both   
4                9.000000                 11          1    both   
5               26.000000                  7          0    both   
6               30.000000                  2          0    both   
7               30.000000                  1          0    both   
8               19.000000                  1          0    both   
9               15.000000                  2          1    both   

     price_range_loc     busiest_day  Busiest_days busiest_period_of_day  \
0  Mid-range product  Regularly busy  Regular days        Average orders   
1  Mid-range product  Regularly busy  Regular days        Average orders   
2  Mid-range product     Busiest day  Busiest days        Average orders   
3  Mid-range product  Regularly busy  Slowest days        Average orders   
4  Mid-range product      Least busy  Slowest days        Average orders   
5  Mid-range product  Regularly busy  Busiest days        Average orders   
6  Mid-range product  Regularly busy  Regular days        Average orders   
7  Mid-range product  Regularly busy  Slowest days           Most orders   
8  Mid-range product  Regularly busy  Busiest days        Average orders   
9  Mid-range product  Regularly busy  Busiest days        Average orders   

  max_order      loyalty_flag  
0        32  Regular customer  
1        32  Regular customer  
2         5      New customer  
3         3      New customer  
4         3      New customer  
5        26  Regular customer  
6         9      New customer  
7        12  Regular customer  
8        20  Regular customer  
9        20  Regular customer  

[10 rows x 21 columns]

In [62]:
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

#  Analyze the result. How do the results for the entire dataframe differ from those of the subset? Include your comments in a markdown cell below the executed code.

All of the mean value for the entire dataframe is totally different from that of the subset. Some were higher, while the others were lower.


#  The team now wants to target different types of spenders in their marketing campaigns.

In [63]:
# Step 1

ords_prods_merge['Ave_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [64]:
ords_prods_merge.head(1)

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id eval_set  order_number  orders_day_of_week  ...  \
0   3139998      138    prior            28                   6  ...   

   add_to_cart_order  reordered  _merge    price_range_loc     busiest_day  \
0                  5          0    both  Mid-range product  Regularly busy   

   Busiest_days busiest_period_of_day max_order      loyalty_flag  Ave_price  
0  Regular days        Average orders        32  Regular customer   6.935811  

[1 rows x 22 columns]

In [65]:
# Step 2

ords_prods_merge.loc[ords_prods_merge['Ave_price'] >=10, 'Spender_pattern'] = 'High spender'


In [66]:
ords_prods_merge.loc[ords_prods_merge['Ave_price'] < 10, 'Spender_pattern'] = 'Low spender'

In [67]:
# Spender_pattern frequency

ords_prods_merge['Spender_pattern'].value_counts(dropna = False)

Low spender     31770614
High spender      634245
Name: Spender_pattern, dtype: int64

There are more Low spenders than High spender.   

In [68]:
# Step 3
ords_prods_merge[['user_id', 'Spender_pattern', 'prices', 'Ave_price']].head(10)

user_id Spender_pattern  prices  Ave_price
0      138     Low spender     5.8   6.935811
1      138     Low spender     5.8   6.935811
2      709     Low spender     5.8   7.930208
3      764     Low spender     5.8   4.972414
4      764     Low spender     5.8   4.972414
5      777     Low spender     5.8   6.935398
6      825     Low spender     5.8   5.957576
7      910     Low spender     5.8   6.680000
8     1052     Low spender     5.8   7.162500
9     1052     Low spender     5.8   7.162500

#  Determine frequent versus non-frequent customers

In [69]:
# Step 1

ords_prods_merge['median_regularity'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [70]:
ords_prods_merge['days_since_prior_order'].head(10)

0     3.000000
1    20.000000
2     6.000000
3     1.111484
4     9.000000
5    26.000000
6    30.000000
7    30.000000
8    19.000000
9    15.000000
Name: days_since_prior_order, dtype: float64

In [71]:
# Step 2

ords_prods_merge.loc[ords_prods_merge['median_regularity'] > 20, 'user_regularity'] = 'Non-frequent customer'

In [72]:
ords_prods_merge.loc[(ords_prods_merge['median_regularity'] <= 20) & (ords_prods_merge['median_regularity'] > 10), 'user_regularity'] = 'Regular customer'

In [73]:
ords_prods_merge.loc[ords_prods_merge['median_regularity'] <= 10, 'user_regularity'] = 'Frequent customer'

In [74]:
# user_regularity frequency

ords_prods_merge['user_regularity'].value_counts(dropna = False)

Frequent customer        22795753
Regular customer          6922378
Non-frequent customer     2686728
Name: user_regularity, dtype: int64

There are more Frequent customer, followed by Regular customer and lastly by Non-frequent customer.

In [75]:
# Step 3

ords_prods_merge[['user_id', 'user_regularity', 'days_since_prior_order', 'median_regularity']].head(10)

user_id    user_regularity  days_since_prior_order  median_regularity
0      138  Frequent customer                3.000000           8.000000
1      138  Frequent customer               20.000000           8.000000
2      709  Frequent customer                6.000000           6.000000
3      764  Frequent customer                1.111484           9.000000
4      764  Frequent customer                9.000000           9.000000
5      777   Regular customer               26.000000          11.000000
6      825   Regular customer               30.000000          19.000000
7      910  Frequent customer               30.000000           3.555742
8     1052  Frequent customer               19.000000           9.500000
9     1052  Frequent customer               15.000000           9.500000

# comparison with another column


In [76]:
# Create a subset

df_new= ords_prods_merge[:1000000]

In [77]:
df_new.shape

(1000000, 25)

In [78]:
df_new.head(10)

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   
5           1  Chocolate Sandwich Cookies        61             19     5.8   
6           1  Chocolate Sandwich Cookies        61             19     5.8   
7           1  Chocolate Sandwich Cookies        61             19     5.8   
8           1  Chocolate Sandwich Cookies        61             19     5.8   
9           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id eval_set  order_number  orders_day_of_week  ...  \
0   3139998      138    prior            28                   6  ...   
1   1977647      138    prior            30                   6  ...   
2    389851      709    prior             2                   0  ...   
3    652770      764    prior             1                   3  ...   
4   1813452      764    prior             3                   4  ...   
5   1701441      777    prior            16                   1  ...   
6   1871483      825    prior             3                   2  ...   
7   1290456      910    prior            12                   3  ...   
8    369558     1052    prior            10                   1  ...   
9    589712     1052    prior            15                   1  ...   

     price_range_loc     busiest_day  Busiest_days  busiest_period_of_day  \
0  Mid-range product  Regularly busy  Regular days         Average orders   
1  Mid-range product  Regularly busy  Regular days         Average orders   
2  Mid-range product     Busiest day  Busiest days         Average orders   
3  Mid-range product  Regularly busy  Slowest days         Average orders   
4  Mid-range product      Least busy  Slowest days         Average orders   
5  Mid-range product  Regularly busy  Busiest days         Average orders   
6  Mid-range product  Regularly busy  Regular days         Average orders   
7  Mid-range product  Regularly busy  Slowest days            Most orders   
8  Mid-range product  Regularly busy  Busiest days         Average orders   
9  Mid-range product  Regularly busy  Busiest days         Average orders   

  max_order      loyalty_flag Ave_price Spender_pattern median_regularity  \
0        32  Regular customer  6.935811     Low spender          8.000000   
1        32  Regular customer  6.935811     Low spender          8.000000   
2         5      New customer  7.930208     Low spender          6.000000   
3         3      New customer  4.972414     Low spender          9.000000   
4         3      New customer  4.972414     Low spender          9.000000   
5        26  Regular customer  6.935398     Low spender         11.000000   
6         9      New customer  5.957576     Low spender         19.000000   
7        12  Regular customer  6.680000     Low spender          3.555742   
8        20  Regular customer  7.162500     Low spender          9.500000   
9        20  Regular customer  7.162500     Low spender          9.500000   

     user_regularity  
0  Frequent customer  
1  Frequent customer  
2  Frequent customer  
3  Frequent customer  
4  Frequent customer  
5   Regular customer  
6   Regular customer  
7  Frequent customer  
8  Frequent customer  
9  Frequent customer  

[10 rows x 25 columns]

In [79]:
df_new.groupby('department_id').agg({'prices': ['mean']})

prices
                    mean
department_id           
1               6.432241
2              11.645328
3               8.952648
4               7.248330
5               5.263945
6               5.943932
7               7.943870
8               8.612695
9               6.366587
10             12.846677
11              8.655505
12             16.578760
13              8.567672
14              9.146213
15              7.500153
16              8.459573
17              9.990755
18              7.864432
19              4.181057
20              8.104981
21              9.105693

In [80]:
# For the entire dataframe

ords_prods_merge.groupby('department_id').agg({'prices': ['mean']}) 

prices
                    mean
department_id           
1               7.736553
2               6.990934
3               7.853809
4               7.981708
5               8.143701
6               7.682264
7               7.680502
8               7.889070
9               7.350283
10              8.348960
11              7.996769
12             16.294443
13              8.009443
14              8.028963
15              7.539976
16             33.479533
17              7.384058
18              7.638380
19              4.275437
20              7.778701
21              8.660716

The analysis of the 2 set indicates that the mean value for the entire dataframe is totally different from that of the subset. Some were higher, while the others were lower.

#  Explore more deeply the difference in the descriptive statistics across the customer groups (i.e. difference in mean, max, median)


In [81]:
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max' , 'median']})

prices                     
                       mean  min      max median
loyalty_flag                                    
Loyal customer    10.386336  1.0  99999.0    7.4
New customer      13.294670  1.0  99999.0    7.4
Regular customer  12.495717  1.0  99999.0    7.4

One one hand, the aggregate average price for loyal customers are lowest at 10.386, followed by the average price for regular customers, while the highest average price is for new customers which is 13.294. This means the new customers are paying more than the loyal and regular customers. On the other hand, the minimum, maximum and median for all groups of customers are the same at 1.0, 99999.0 and 7.4.

In [82]:
# Turning outliers to NaNs

ords_prods_merge.loc[ords_prods_merge['prices'] >100, 'prices'] = np.nan

# Export the newly created dataframe as orders_products_merged

In [83]:
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data for storing any manipulated data sets', 'ords_prods_new columns.pkl'))

In [84]:
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id eval_set  order_number  orders_day_of_week  ...  \
0   3139998      138    prior            28                   6  ...   
1   1977647      138    prior            30                   6  ...   
2    389851      709    prior             2                   0  ...   
3    652770      764    prior             1                   3  ...   
4   1813452      764    prior             3                   4  ...   

     price_range_loc     busiest_day  Busiest_days  busiest_period_of_day  \
0  Mid-range product  Regularly busy  Regular days         Average orders   
1  Mid-range product  Regularly busy  Regular days         Average orders   
2  Mid-range product     Busiest day  Busiest days         Average orders   
3  Mid-range product  Regularly busy  Slowest days         Average orders   
4  Mid-range product      Least busy  Slowest days         Average orders   

  max_order      loyalty_flag Ave_price Spender_pattern median_regularity  \
0        32  Regular customer  6.935811     Low spender               8.0   
1        32  Regular customer  6.935811     Low spender               8.0   
2         5      New customer  7.930208     Low spender               6.0   
3         3      New customer  4.972414     Low spender               9.0   
4         3      New customer  4.972414     Low spender               9.0   

     user_regularity  
0  Frequent customer  
1  Frequent customer  
2  Frequent customer  
3  Frequent customer  
4  Frequent customer  

[5 rows x 25 columns]